### ISSUE TITLE
Find appropriate Data Augmentation Script and Apply it to the dataset (subset of dataset)

### ISSUE THREAD

**Issue by AntonDahl** [rverschuren/STPM-Unet-for-industrial-anomaly-detection#2]
Find some scripts for dataAug here: https://github.com/AgaMiko/data-augmentation-review/blob/master/README.md#Computer-vision

**Comment by rverschuren** [comment]
The interest of performing data augmentation for this project is double:

- Improving the performances of the model.
- Creating model abstraction of the colors and the textile density.

One way of measuring the second point is to apply anomaly detection on another type of textile and see how it can perform.

It is important to choose the right augmentation operations to avoid creating anomalies by augmenting the data. 

Some ideas to implement
- [ ] Blurring: to simulate different densities of textile. 
- [ ] Colors
- [ ] others...

If you want to run the model and output the augmentation to observe what it is doing: 

Run the following command
`!python train_resnet.py --phase augment --category carpet --num_epochs 3`

**Comment by rverschuren** [comment]
Using this library : [https://github.com/kornia/kornia](https://github.com/kornia/kornia)

And this example https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/augmentation_kornia.html

**Comment by rverschuren** [comment]
### Default
**Code:**
```python
        self.data_transforms = transforms.Compose([
                        transforms.Resize((args.load_size, args.load_size), transforms.InterpolationMode.LANCZOS),
                        transforms.ToTensor(),
                        transforms.CenterCrop(args.input_size),
                        transforms.Normalize(mean=mean_train,
                                            std=std_train)])
```
**Image:**
![image](https://user-images.githubusercontent.com/23149720/205429182-dee84c69-9930-4aa1-a8f5-6a15edc8a749.png)


### Default mode but without normalization and centercrop.
**Code:**
```python
        self.data_transforms = transforms.Compose([
                        transforms.Resize((args.load_size, args.load_size), transforms.InterpolationMode.LANCZOS),
                        transforms.ToTensor()    ])
```
**Image:**
![image](https://user-images.githubusercontent.com/23149720/205429196-65f79980-e9fd-4a50-9ab9-4efbd1124dfe.png)

**Comment by rverschuren** [comment]
If you want to add some augmentations.

1. Import it by adding the class in the import line
`from kornia.augmentation import RandomBoxBlur, ColorJitter, Normalize [OTHER.....]`

2. Add in the sequential composition of augmentations:
```python
class DataAugmentation(nn.Module):
    """Module to perform data augmentation using Kornia on torch tensors."""

    def __init__(self, apply_color_jitter: bool = False) -> None:
        super().__init__()
        self._apply_color_jitter = apply_color_jitter

        self.transforms = nn.Sequential(
            RandomBoxBlur(kernel_size=(3, 3), border_type='reflect', p=0.75),
            Normalize(mean=mean_train, std=std_train)
            [ADD HERE]
        )
```

3. Run to observe output files:
`!python train_resnet.py --phase augment --category carpet --num_epochs 3`

4. open the 2 generate files. 
> "img_augment.png"
> "img_before_augment.png"

**Comment by rverschuren** [comment]
I tried to add the RandomRotation, but the problem is that we lose information on the image and a part of the image is set to black pixels. Which could be detected as an anomaly. By using the `RandomAffine(degrees=45.0, scale=(1,2), padding_mode=2, p=.75)`, we can set the scale value and the padding_mode which allows us to keep the texture of the textile everywhere on the image. 

Se the difference of RandomRotation and RandomAffine on the newt two images.

![Screenshot from 2022-12-03 15-54-22](https://user-images.githubusercontent.com/23149720/205429069-27d307d6-a2ee-4a4d-98d7-1a75aa34e136.png)

![img_augment (3)](https://user-images.githubusercontent.com/23149720/205429092-ddf719f0-c08c-434c-8e1f-2213c7f7a3fc.png)

**Comment by rverschuren** [comment]
Now let's compare when playing with colors, contrast, brightness, ...

Using the kornia.augmentation.ColorJiggle class.

To better understand the difference, I did not normalize the data before saving the image.
The two next images are 
- with brightness=0.1, contrast=0.1, saturation=0.1, hue=0.1

> Without normalization
![jiggle01](https://user-images.githubusercontent.com/23149720/205430197-155e846a-2c18-4cd8-bf08-6bd7d4bf3ffe.png)
> With normalization
![img_augment (5)](https://user-images.githubusercontent.com/23149720/205430533-95cf1684-0858-46e0-9086-c28c7f38b45b.png)


- with brightness=0.2, contrast=0.2, saturation=0.2, hue=0.2
> Without normalization
![jiggle02](https://user-images.githubusercontent.com/23149720/205430205-035e0805-4c04-4128-8643-d64aeb169dc4.png)
> With normalization
![img_augment (4)](https://user-images.githubusercontent.com/23149720/205430393-b08698e3-8546-4432-84e7-6f8bd07591b0.png)


In the commit I set it to 0.1

**Comment by rverschuren** [comment]
https://www.mydrive.ch/shares/38536/3830184030e49fe74747669442f0f282/download/420937607-1629951964/leather.tar.xz

**Comment by rverschuren** [comment]
Adding some blur to obtain something similar to leather.

```python
        self.transforms = nn.Sequential(
            RandomBoxBlur(kernel_size=(2,2), border_type='reflect', p=0.2),
            RandomBoxBlur(kernel_size=(3,3), border_type='reflect', p=0.2),
            RandomBoxBlur(kernel_size=(5,5), border_type='reflect', p=0.2),
            RandomBoxBlur(kernel_size=(7,7), border_type='reflect', p=0.2),
            RandomBoxBlur(kernel_size=(5,5), border_type='reflect', p=0.2),
            RandomBoxBlur(kernel_size=(15,15), border_type='reflect', p=0.1),
            RandomAffine(degrees=45.0, scale=(1,2), padding_mode=2, p=.75),
            ColorJiggle(0.1, 0.1, 0.1, 0.1, p=1.),
            #Normalize(mean=mean_train, std=std_train)
        )
```
![img_augment (6)](https://user-images.githubusercontent.com/23149720/205431925-fb6c1083-bf32-4a62-b29a-e9491cdbd542.png)

![img_augment (7)](https://user-images.githubusercontent.com/23149720/205432113-3b00909d-89f3-4a3d-a31b-f05c0d447213.png)
